[inductor] Add an AOT mode for the Triton backend

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98088
* __->__ #97152

Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

Differential Revision: [D44407904](https://our.internmc.facebook.com/intern/diff/D44407904)

desertfire committed:
[inductor] Add an AOT mode for the Triton backend

Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Add an AOT mode for the Triton backend"


Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire commented:
Generated code example,
```
/* AOT-Inductor generated code */

#include <ATen/ATen.h>
#include <ATen/cuda/CUDAContext.h>
#include <ATen/cuda/Exceptions.h>
#include <ATen/cuda/nvrtc_stub/ATenNVRTC.h>

#include <fstream>

CUfunction loadKernel(const std::string &filePath, const std::string &funcName) {
    CUmodule mod;
    CUfunction func;
    AT_CUDA_DRIVER_CHECK(cuModuleLoad(&mod, filePath.c_str()));
    AT_CUDA_DRIVER_CHECK(cuModuleGetFunction(&func, mod, funcName.c_str()));
    return func;
}

void launchKernel(
        CUfunction func,
        int gridX,
        int gridY,
        int gridZ,
        int numWraps,
        int sharedMemBytes,
        void* args[]) {
    AT_CUDA_DRIVER_CHECK(cuLaunchKernel(
        func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes,
        at::cuda::getCurrentCUDAStream(), args, nullptr));
}

static CUfunction triton_fused_add_0 = nullptr;

at::Tensor aot_inductor_entry(std::vector<at::Tensor> args) {
    at::Tensor arg0_1, arg1_1;
    arg0_1 = args[0];
    arg1_1 = args[1];
    auto buf0 = at::empty_strided({32, 64}, {64, 1}, at::device(at::kCUDA).dtype(at::ScalarType::Float));
    if (triton_fused_add_0 == nullptr) {
         triton_fused_add_0 = loadKernel("triton_fused_add_0.cubin", "triton_fused_add_0_0d1d2d3d");
    }
    CUdeviceptr var_0 = reinterpret_cast<CUdeviceptr>(arg1_1.data_ptr());
    CUdeviceptr var_1 = reinterpret_cast<CUdeviceptr>(arg0_1.data_ptr());
    CUdeviceptr var_2 = reinterpret_cast<CUdeviceptr>(buf0.data_ptr());
    int var_3 = 2048;
    void* kernel_args_0[] = {&var_0, &var_1, &var_2, &var_3};
    launchKernel(triton_fused_add_0, 8, 1, 1, 1, 0, kernel_args_0);
    arg0_1.reset();
    arg1_1.reset();
    return buf0;
}
```

jansel commented on torch/_inductor/codecache.py:
```diff
@@ -447,7 +447,9 @@ def cpp_flags():
-def optimization_flags():
+def optimization_flags(cuda=False):
+    if cuda:
+        return "-O0 -g"
```
What is this?

jansel commented on torch/_inductor/codecache.py:
```diff
@@ -567,11 +581,14 @@ def compile(cls, source_code):
+                        vec_isa=picked_vec_isa,
+                        cuda=cuda,
                     ).split(" ")
+                    log.info(" ".join(cmd))
```
Remove?

jansel commented on torch/_inductor/graph.py:
```diff
@@ -532,12 +538,17 @@ def check_profiler_mark_wrapper_call(self):
-                return
-        self.disable_cpp_wrapper("device not CPU")
+            device = list(self.device_types)[0]
+            if device == "cpu" or "cuda":
```
This always returns `True`

ngimel commented:
I don't think cpp wrapper properly generates device guards and stream calls that python wrapper is generating?

desertfire committed:
Update on "[inductor] Add an AOT mode for the Triton backend"


Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Add an AOT mode for the Triton backend"


Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -150,6 +173,38 @@ def codegen(self, code: IndentedBuffer):
+        self.name = name
+
+    def __call__(self):
+        with open(f"{self.name}.mangled", "r") as f:
```
If we need tempfile here and below (I dislike the tempfile thing):
1) We should make sure it is threadsafe if multiple PyTorch procs are running -- I don't think this is
2) We should put the tempfiles in the torchindutor cache dir not the working dir.

jansel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -897,10 +954,163 @@ def wrap_kernel_call(self, name, call_args):
+        self.kernel_callsite_id += 1
+
+
+class TwoPassAotWrapperCodeGen(WrapperCodeGen):
```
This two pass approach feels a bit clunky when implemented here.  I wonder if we could simplify things by moving the two passes up to the caller's level.

Basically, require the user to do a warmup run.  Or have a wrapper function outside inductor that does the warmup run via two calls to `compile_fx`.

When compiling in AOT mode, we can basically just assert that everything is a cache hit, so all the autotuning will be a no-op.

desertfire replied:
Requiring user to do an explicit warmup run sounds a little award to me, although I agree calling `compile_fx` twice makes the code cleaner and more robust. I plan to change `aot_compile` in `_inductor/__init__.py` to achieve that. How does that sound?

jansel commented on torch/_inductor/triton_ops/autotune.py:
```diff
@@ -193,6 +198,22 @@ def run(self, *args, grid, stream):
+                f.write(
+                    f"{grid_0}, {grid_1}, {grid_2}, {launcher.bin.num_warps}, {launcher.bin.shared}"
+                )
+            with open(f"{launcher.kernel_name}.cubin", "wb") as f:
```
Not safe for multiprocessing, leaks files into working dir.

desertfire committed:
Update on "[inductor] Add an AOT mode for the Triton backend"


Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Add an AOT mode for the Triton backend"


Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44407904).

ngimel commented:
> > I don't think cpp wrapper properly generates device guards and stream calls that python wrapper is generating?
> 
> I have updated the generated code example to reflect the change I made. Basically it's calling `at::cuda::CUDAGuard device_guard(device_index);` to guard the device and then`at::cuda::getCurrentCUDAStream()` to get the stream. Let me know if that is sufficient.

Calling deviceGuard for every kernel launch is wasteful, we should follow the same guard strategy for device guards that python codegen is using - wrap code in device guard, and create a new device guard (or reset the device) only when device is actually changing.

desertfire committed:
Update on "[inductor] Add an AOT mode for the Triton backend"


Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

Differential Revision: [D44407904](https://our.internmc.facebook.com/intern/diff/D44407904)

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Add an AOT mode for the Triton backend"


Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

Differential Revision: [D44407904](https://our.internmc.facebook.com/intern/diff/D44407904)

[ghstack-poisoned]

desertfire commented:
@jansel , I believe I have addressed all your previous comments. Currently the concepts of cpp_wrapper and aot_mode are tangled. I would like to take another refactoring after this one lands.

ngimel commented on torch/_inductor/graph.py:
```diff
@@ -649,6 +666,8 @@ def compile_to_module(self):
         from .codecache import PyCodeCache
 
         code, linemap = self.codegen()
+        if config.debug:
```
please remove this, or the generated code will be printed twice (it is printed on line 677)
Also, use standard logging facilities in `compile_to_fn` instead of `if config.debug: print...`

ngimel commented on torch/_inductor/compile_fx.py:
```diff
@@ -451,6 +455,75 @@ def is_not_gradout(x):
+                }
+            )
+            with config.patch(new_config_patches):
+                module_copy = deepcopy(module)
```
will this lead to greatly increased memory usage?

desertfire committed:
Update on "[inductor] Add an AOT mode for the Triton backend"


Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

Differential Revision: [D44407904](https://our.internmc.facebook.com/intern/diff/D44407904)

[ghstack-poisoned]

ngimel commented on torch/_inductor/graph.py:
```diff
@@ -666,13 +682,14 @@ def compile_to_fn(self):
-                print(code)
+            code, linemap = self.codegen()
+            log.debug("/* AOT-Inductor generated code */")
+            log.debug(code)
```
```suggestion
            output_code_log.debug(code)
```

desertfire committed:
Update on "[inductor] Add an AOT mode for the Triton backend"


Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

Differential Revision: [D44407904](https://our.internmc.facebook.com/intern/diff/D44407904)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44407904).

desertfire committed:
Update on "[inductor] Add an AOT mode for the Triton backend"


Summary: This PR implements a two-pass wrapper codegen for the Triton
backend to achieve ahead-of-time compilation. In the first pass, the
regular python wrapper code will be generated, and then the generated
code will be executed to perform Triton compilation and autotuning.
After that, the second pass wrapper codegen will generate C++ wrapper
with proper CUDA API to load and launch Triton-generated CUDA kernels.

Like the AOT mode for the cpp backend, the next step would be to provide
a more complete API for AOT.

Differential Revision: [D44407904](https://our.internmc.facebook.com/intern/diff/D44407904)

[ghstack-poisoned]

desertfire commented:
Landed with https://github.com/pytorch/pytorch/pull/98214
